### Recursion

```
error NegativeFactorialInput ; or something

factorial = (x: Integer) -> Integer {
    if x < 0 {
        panic NegativeFactorialInput
    }
    if x == 0 {
        return 1
    }
    return self(x - 1) * x
}
```

Since all functions are anonymous, we can't access a function by its name before it's actually assigned to it.\
(Sometimes we don't assign functions at all, just use them in-place, what then?)\
We need `self` keyword, simple. Kind of like `this` for objects.

By the way 1: you got a little preview tease of errors ("exceptions").\
By the way 2: this is not the best way (or at least only way) to write factorial. Check below.

### Tail

To ensure your function is tail-recursive and therefore optimized by the compiler, use the `tail` keyword:

```
fact = tail (argument: UnsignedInteger, accumulator: UnsignedInteger) -> UnsignedInteger {
    if argument == 0 {
        return 1
    }
    return self(argument - 1, accumulator * argument)
}

export factorial = (x: UnsignedInteger) -> UnsignedInteger {
    return fact(x, 1)
}
```

If your function can't be tail-optimized, you'll get a nice error, preventing you from walking away proud of a job well done, when it's not.

Also, I guess the compiler may still decide to tail-optimize if it's possible, even without `tail`.

By the way 1: Avoid errors, like here, by declaring proper types (see chapter 9).\
By the way 2: I'd keep `fact` private, because someone, somewhere, sooner or later would use it with accumulator other than 1.